src/core/variable_mx.jl

```diff
@@ -0,0 +1,529 @@
+# + or no bounds at all.
+
+# General matrix variables of size (N,M) are created by
+# + variable_mx_real
+# + variable_mx_complex
+# An alternative is provided which allows to pass the diagonal elements as
+# a keyword argument; this is a use case that appears often for P,Q matrices
+# + variable_mx_real_with_diag
+# + variable_mx_complex_with_diag
+
+# Symmetric, skewsymmetric and Hermitian matrix variables are always square
+# variable_mx_symmetric
+# variable_mx_skewsymmetric
+# variable_mx_hermitian
+
+# A special function is available for the Hermitian matrix of the form W=UU',
+# allowing to pass magnitude bounds on U instead
+# variable_mx_hermitian_sqrt_bounds
+
+# Most of the method arguments are strongly typed, to make it clear what is expected.
+# These functions only depend on JuMP models, not PowerModels.
+
+
+# GENERAL MATRICES
+#%##########################
+
+
+"""
+This function creates a set of real matrix variables of size NxM,
+indexed over the elements of the indices argument.
+The upper and lower bounds have to be specified,
+and are dictionaries with the indices as keys and the matrix bounds as values.
+The name and prefix arguments will be combined into the base_name argument for
+JuMP; the prefix will typically be the network number nw.
+Instead of sequentially creating the matrix variables, the elements of the
+matrices are created sequentially for all matrices at once. I.e., we loop
+over the elements, and not over the indices. This is needed so that the
+variable names printed by JuMP are in line with the current design.
+
+Returns a dictionary of (index, matrix  variable) pairs
+"""
+function variable_mx_real(model::JuMP.Model, indices::Array{T,1}, N::Int, M::Int,
+        upper_bound::Dict{T,Array{UB,2}}, lower_bound::Dict{T,Array{LB,2}};
+        name="", prefix="") where {T, LB<:Real, UB<:Real}
+    # the output is a dictionary of (index, matrix) pairs
+    dict_mat_vars_temp = Dict{T, Array{Any, 2}}([(i,fill(missing,N,M)) for i in indices])
+    for n in 1:N
+        for m in 1:M
+            varname = isempty(prefix) ? "$(n)$(m)_$name" : "$(prefix)_$(n)$(m)_$name"
+            # create the element (n,m) for all indices
+            mat_nm = JuMP.@variable(model, [index in indices],
+                base_name=varname,
+                lower_bound=lower_bound[index][n,m],
+                upper_bound=upper_bound[index][n,m]
+            )
+            # unpack element (n,m) to the correct place in the ouput dict
+            for index in indices
+                dict_mat_vars_temp[index][n,m] = mat_nm[index]
+            end
+        end
+    end
+    dict_mat_vars = Dict{T, Array{JuMP.VariableRef, 2}}([(id, arr) for (id, arr) in dict_mat_vars_temp])
+
+    return dict_mat_vars
+end
+
+
+"""
+Alternative method of variable_mx_real,
+that does not take any bounds, and creates unbounded matrix variables.
+"""
+function variable_mx_real(model::JuMP.Model, indices::Array{T,1}, N::Int, M::Int;
```

Comment:
Would it be possible to combine the unbounded versions of functions with the bounded ones instead of overloading them?